[c10d][fr] Make FR vendor neutral so that other backends can use it

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #152614
* __->__ #152563
* #152585

Current FR code is built with `USE_C10D_NCCL` we should remove it to make it generic. And we keep existing API used by NCCL so that we can have some bc compatibility because lots of use cases are around FR with NCCL. The generic version with C10::Event can then be used for other backend like Gloo, etc.

The current Unit test should cover the change.

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @d4l3kC

fduwjj committed:
[c10d][fr] Make FR vendor neutral so that other backends can use it

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Make FR vendor neutral so that other backends can use it"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Make FR vendor neutral so that other backends can use it"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Make FR vendor neutral so that other backends can use it"


Current FR code is built with `USE_C10D_NCCL` we should remove it to make it generic. And we create a special child struct for NCCL to swap c10::Event with CudaEvent so that we can have some bc compatibility because lots of use cases are around FR with NCCL. The generic version with C10::Event can then be used for other backend like Gloo, etc.

The current Unit test should cover the change.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3kC

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Make FR vendor neutral so that other backends can use it"


Current FR code is built with `USE_C10D_NCCL` we should remove it to make it generic. And we create a special child struct for NCCL to swap c10::Event with CudaEvent so that we can have some bc compatibility because lots of use cases are around FR with NCCL. The generic version with C10::Event can then be used for other backend like Gloo, etc.

The current Unit test should cover the change.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3kC

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Make FR vendor neutral so that other backends can use it"


Current FR code is built with `USE_C10D_NCCL` we should remove it to make it generic. And we create a special child struct for NCCL to swap c10::Event with CudaEvent so that we can have some bc compatibility because lots of use cases are around FR with NCCL. The generic version with C10::Event can then be used for other backend like Gloo, etc.

The current Unit test should cover the change.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3kC

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Make FR vendor neutral so that other backends can use it"


Current FR code is built with `USE_C10D_NCCL` we should remove it to make it generic. And we create a special child struct for NCCL to swap c10::Event with CudaEvent so that we can have some bc compatibility because lots of use cases are around FR with NCCL. The generic version with C10::Event can then be used for other backend like Gloo, etc.

The current Unit test should cover the change.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3kC

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Make FR vendor neutral so that other backends can use it"


Current FR code is built with `USE_C10D_NCCL` we should remove it to make it generic. And we keep existing API used by NCCL so that we can have some bc compatibility because lots of use cases are around FR with NCCL. The generic version with C10::Event can then be used for other backend like Gloo, etc.

The current Unit test should cover the change.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3kC

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Make FR vendor neutral so that other backends can use it"


Current FR code is built with `USE_C10D_NCCL` we should remove it to make it generic. And we keep existing API used by NCCL so that we can have some bc compatibility because lots of use cases are around FR with NCCL. The generic version with C10::Event can then be used for other backend like Gloo, etc.

The current Unit test should cover the change.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3kC

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.hpp:
```diff
@@ -1,16 +1,14 @@
 #include <mutex>
 
 #include <ATen/ATen.h>
+#ifdef USE_C10D_NCCL
 #include <ATen/cuda/CUDAEvent.h>
+#endif // USE_C10D_NCCL
```
The only place that uses `CUDAEvent` is `getDurationFromEvent` function. Shall we template this function and remove `#include <ATen/cuda/CUDAEvent.h>` entirely?

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.hpp:
```diff
@@ -89,10 +87,15 @@ class TORCH_API DebugInfoWriter {
+/* Helper used by work::getDuration() and flight recorder */
+float getDurationFromEvent(c10::Event& startEvent, c10::Event& endEvent);
+
+#ifdef USE_C10D_NCCL
 /* Helper used by work::getDuration() and nccl flight recorder */
 float getDurationFromEvent(
     at::cuda::CUDAEvent& ncclStartEvent,
     at::cuda::CUDAEvent& ncclEndEvent);
+#endif // USE_C10D_NCCL
```
See comment above

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.hpp:
```diff
@@ -242,23 +247,37 @@ struct FlightRecorder {
+      bool onlyActive);
 
-#endif // USE_C10D_NCCL
+  using json = nlohmann::json;
+  json get_dump_json(bool includeCollectives, bool onlyActive);
```
nit: this is a private function -- no one is going to use from outside.
Let see if there is a way to hide it from the class's definition, so that we don't need to add `#include <nlohmann/json.hpp>` in the header -- better to have it in cpp file.

d4l3k replied:
Why'd we have to change the signature on this to return json instead of str? If we keep it as string we can keep include in cpp file iiuc

fduwjj replied:
oh I just want to extract the common logic into a private util functio.

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.hpp:
```diff
@@ -242,23 +247,37 @@ struct FlightRecorder {
+
+#ifdef USE_C10D_NCCL
   std::string dump_json(
       const std::optional<std::unordered_map<
           std::string,
           std::unordered_map<std::string, std::string>>>& ncclDumpMap,
       bool includeCollectives,
       bool onlyActive);
 
-  // dump all collectives + ncclDumpMap
   std::string dump(
       const std::optional<std::unordered_map<
           std::string,
           std::unordered_map<std::string, std::string>>>& ncclDumpMap,
       bool includeCollectives,
       bool includeStackTraces,
       bool onlyActive);
-};
+#endif // USE_C10D_NCCL
```
No need for `USE_C10D_NCCL` if these functions can compile independently.

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -1,21 +1,23 @@
-// TODO: Make Fligth Recorder device agnostic
 #ifdef USE_C10D_NCCL
```
Why do we still have `#ifdef USE_C10D_NCCL` here?

fduwjj replied:
because #include <cuda_runtime.h>

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -678,6 +670,58 @@ std::string FlightRecorder<EventType>::dump(
+std::unique_ptr<DebugInfoWriter> DebugInfoWriter::writer_ = nullptr;
+std::atomic<bool> DebugInfoWriter::hasWriterRegistered_(false);
+
+float getDurationFromEvent(c10::Event& startEvent, c10::Event& endEvent) {
+  TORCH_CHECK(false, "getDuration not supported by c10::Event.");
+}
```
That's a surprise. cc @albanD

fduwjj replied:
now you see my point... I try the way cuda event is doing but I got compile error.

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -678,6 +670,58 @@ std::string FlightRecorder<EventType>::dump(
+  return result;
+}
+
+template <typename EventType>
+std::string FlightRecorder<EventType>::dump_json(
+    bool includeCollectives,
+    bool onlyActive) {
+  return (get_dump_json(includeCollectives, onlyActive)).dump();
+}
+
+template <typename EventType>
+// dump all collectives + ncclDumpMap
+std::string FlightRecorder<EventType>::dump(
+    bool includeCollectives,
+    bool includeStackTraces,
+    bool onlyActive) {
+  return pickle_str(
+      get_dump(includeCollectives, includeStackTraces, onlyActive));
+}
```
nit: see if you could simplify the wrapping? (not an issue of this PR)

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -678,6 +670,58 @@ std::string FlightRecorder<EventType>::dump(
+  TORCH_CHECK(false, "getDuration not supported by c10::Event.");
+}
+
+#ifdef USE_C10D_NCCL
+template <typename EventType>
+std::string FlightRecorder<EventType>::dump_json(
+    const std::optional<std::unordered_map<
+        std::string,
+        std::unordered_map<std::string, std::string>>>& ncclDumpMap,
+    bool includeCollectives,
+    bool onlyActive) {
+  // Adding ncclDumpMap to the json object is NCCL specific.
+  auto result = get_dump_json(includeCollectives, onlyActive);
+  if (ncclDumpMap.has_value()) {
+    result[nccl_comm_key_str] = ncclDumpMap.value();
+  }
+  return result.dump();
+}
```
nit: which class/variable requires `#ifdef USE_C10D_NCCL` ? `ncclDumpMap`? But it seems to be a generic type: `std::unordered_map<std::string, std::string>>>`

fduwjj replied:
currently ncclDumpMap is mostly used by ncclx, if we believe this is necessary, I won't remove it.

kwen2501 reviewed (approved):
Overall looks good to me. Wish we could remove more `#ifdef USE_C10D_NCCL` (see comments)

d4l3k commented on torch/csrc/distributed/c10d/FlightRecorder.hpp:
```diff
@@ -104,8 +107,10 @@ struct FlightRecorder {
+    max_entries_ = getCvarInt(
+        {"TORCH_TRACE_BUFFER_SIZE", "TORCH_NCCL_TRACE_BUFFER_SIZE"}, 0);
+    capture_cpp_stack_ = getCvarBool(
+        {"TORCH_TRACE_CPP_STACK", "TORCH_NCCL_TRACE_CPP_STACK"}, false);
```
nit: maybe we should name this `TORCH_FR_*` or `TORCH_C10D_TRACE_.*`?

Trace is a bit overloaded since it could also be referring to profilers etc

d4l3k reviewed (approved):
LGTM, agree with @kwen2501 comments

fduwjj commented:
I wish to remove but there are still some cuda related logics inside `dumpHandler` and also c10::Event does not have the same behavior as cudaEvent. So we just keep them for now.

fduwjj committed:
Update on "[c10d][fr] Make FR vendor neutral so that other backends can use it"


Current FR code is built with `USE_C10D_NCCL` we should remove it to make it generic. And we keep existing API used by NCCL so that we can have some bc compatibility because lots of use cases are around FR with NCCL. The generic version with C10::Event can then be used for other backend like Gloo, etc.

The current Unit test should cover the change.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3kC

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][fr] Make FR vendor neutral so that other backends can use it"


Current FR code is built with `USE_C10D_NCCL` we should remove it to make it generic. And we keep existing API used by NCCL so that we can have some bc compatibility because lots of use cases are around FR with NCCL. The generic version with C10::Event can then be used for other backend like Gloo, etc.

The current Unit test should cover the change.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3kC

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj committed:
Update on "[c10d][fr] Make FR vendor neutral so that other backends can use it"


Current FR code is built with `USE_C10D_NCCL` we should remove it to make it generic. And we keep existing API used by NCCL so that we can have some bc compatibility because lots of use cases are around FR with NCCL. The generic version with C10::Event can then be used for other backend like Gloo, etc.

The current Unit test should cover the change.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k d4l3kC

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

guangyey commented:
FYI @zhangxiaoli73 @frost-intel
